File: mybuffer/my_buffer.cc
#include "./my_buffer.h"

MyBuffer::MyBuffer(int size){
    this->s_ = new std::string(size, char('\0'));
}

MyBuffer::~MyBuffer() {
    delete this->s_;
}

int MyBuffer::Size() const {
    return this->s_->size();
}
char* MyBuffer::Data() {
    return (char*)this->s_->data();
}

void MyBuffer::Push(int value){
    this->vectorInt.push_back(value);
}

int MyBuffer::Pop(){
   return this->vectorInt.front();
}

File: mybuffer/my_buffer.go
// Copyright © 2017 ChaiShushan <chaishushan{AT}gmail.com>.
// License: https://creativecommons.org/licenses/by-nc-sa/4.0/

package mybuffer

import "unsafe"

type MyBuffer struct {
	cptr *cgo_MyBuffer_T
}

func NewMyBuffer(size int) *MyBuffer {
	return &MyBuffer{
		cptr: cgo_NewMyBuffer(size),
	}
}

func (p *MyBuffer) Delete() {
	cgo_DeleteMyBuffer(p.cptr)
}

func (p *MyBuffer) Data() []byte {
	data := cgo_MyBuffer_Data(p.cptr)
	size := cgo_MyBuffer_Size(p.cptr)
	return ((*[1 << 31]byte)(unsafe.Pointer(data)))[0:int(size):int(size)]
}

func (p MyBuffer) Push(value int) {
	cgo_MyBuffer_Push(p.cptr, value)
}

func (p MyBuffer) Pop() int {
	return int(cgo_MyBuffer_Pop(p.cptr))
}
